c2rust-analyze/src/rewrite/span_index.rs

```diff
@@ -0,0 +1,54 @@
+//! Defines the [`SpanIndex`] data structure for looking up MIR statements by span.
+use rustc_span::{BytePos, Span};
+use std::slice;
+
+/// A mapping from [`Span`]s to values of type `T`.  Allows looking up all items whose span is a
+/// subset of some target span.
+pub struct SpanIndex<T> {
+    v: Vec<(Span, T)>,
+}
+
+impl<T> SpanIndex<T> {
+    /// Construct a new [`SpanIndex`], containing the items provided by the iterator `it`.
+    pub fn new(it: impl IntoIterator<Item = (Span, T)>) -> SpanIndex<T> {
+        let mut v = it.into_iter().collect::<Vec<_>>();
+        v.sort_by_key(|(span, _)| span.lo());
+        SpanIndex { v }
+    }
+
+    /// Iterate over items whose spans are entirely contained within `span`.  The order of the
+    /// returned items is unspecified.
+    pub fn lookup<'a>(&'a self, span: Span) -> RangeIter<'a, T> {
+        let data = span.data();
+        let start = self.v.partition_point(|(span, _)| span.lo() < data.lo);
```

Comment:
shouldn't the input param to `lookup`, ,`span`, have `span.lo()` available? you'd need to change the `span` in the `partition_point` closure to another name. I'm unsure why `data = span.data()` is necessary

Reply:
This is a tiny optimization.  Calling `span.lo()` is equivalent to `span.data().lo`, and if `span` is interned, each `span.data()` is somewhat expensive since it accesses the span interner.  With this approach, we only access the interner once, no matter how many times the closure runs.